rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -181,7 +200,29 @@ void Player::play_messages_until_queue_empty(const PlayOptions & options)
   while (message_queue_.try_dequeue(message) && rclcpp::ok()) {
     std::this_thread::sleep_until(
       start_time_ + std::chrono::duration_cast<std::chrono::nanoseconds>(
-        1.0 / rate * message.time_since_start));
+        1.0 / rate * message.time_since_start + paused_duration_));
+
+    if (rosbag2_transport_->get_current_state().id() !=
+      lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE)
```

Comment:
what happened if in state `unconfigured` or `finalized`?

Reply:
With respect to your first comment, the ROS 1 bag implementation had a time_translator class that took care of manipulating the time and keeping track of time pauses and associated math. Do you think this is a better approach?

Reply:
I think the question is referring to https://github.com/ros2/rosbag2/pull/443#discussion_r451010655
@Karsten1987 Is the time_translator way raised above preferred for accounting for the pause duration?

Reply:
> @danthony06 please apologize, your comment flew under my radar. Also, I feel Github renders the chronological order of inline comments weird, so let me just answer here:
> 
> > With respect to your first comment, the ROS 1 bag implementation had a time_translator class that took care of manipulating the time and keeping track of time pauses and associated math. Do you think this is a better approach?
> 
> I think it makes sense to abstract the timing behavior out of the `player` class, which handles things like the duration we've been paused or the ratio in which messages are being replayed. I guess we can also leverage this later on when replaying with simulated time.
> I am generally just concerned whether "sleeping" is the right way of seeking for the next message to be ready, but I guess we can roll this way as for now. How do you envision the `time_translator` class API to look like? Returning a delta for each loop until when we sleep?

Sorry, I'm even tardier in responding to this. I'm working through it, but I was thinking about something like this:
https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag/src/time_translator.cpp
Which would allow us to pass in the current time, and get back the delta for sleeping.